File: leetcode/algorithms/590_n_ary_tree_postorder_traversal/main.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class NAryTreePostorderTraversal:
    def postorder(self, root: "Node") -> List[int]:
        return self.traversal(root, [])

    def traversal(self, node: "Node", list: List[int]) -> List[int]:
        if node == None:
            return list

        for child in node.children:
            self.traversal(child, list)

        list.append(node.val)

        return list


    # Solution
    # Solution 1: Recursion
    def solution1(self, root: "Node") -> List[int]:
        res = []

        if root == None:
            return res

        def recursion(root: "Node", res: List[int]):
            for child in root.children:
                recursion(child, res)
            res.append(root.val)

        recursion(root, res)

        return res

    # Solution 2: Iteration
    def solution2(self, root: "Node") -> List[int]:
        res = []

        if root == None:
            return res

        stack = [root]

        while stack:
            curr = stack.pop()
            res.append(curr.val)
            stack.extend(curr.children)

        return res[::-1]

File: leetcode/algorithms/590_n_ary_tree_postorder_traversal/main.js
function Node(val, children) {
  this.val = val;
  this.children = children;
}

/**
 * @param {Node|null} root
 * @param {number[]} result
 * @return {number[]}
 */
var postorder = function (root, result = []) {
  if (!root) {
    return result;
  }

  for (const child of root.children) {
    postorder(child, result);
  }

  result.push(root.val);

  return result;
};


// Solution
// Solution 1: Iteravtive Post-order Traverse
var solution1 = function (root) {
  const res = [],
    stack = [root];

  while (stack.length) {
    const curr = stack.pop();
    if (!curr) {
      continue;
    }
    res.push(curr.val);
    stack.push(...curr.children);
  }

  return res.reverse();
};

// Solution 2: Recursive Post-order Traverse
var solution2 = function (root) {
  const res = [];

  function traverse(node) {
    if (!node) {
      return;
    }
    for (child of node.children) {
      traverse(child);
    }
    res.push(node.val);
  }

  traverse(root);

  return res;
};
